Source/Methods/Envelope Tracking/EnvelopeTracking.cs

```diff
@@ -0,0 +1,309 @@
+            scaledEnvelopeWaveform.Data = envelopeWaveform.Data.Clone();
+            WritableBuffer<ComplexSingle> scaledEnvelopeWaveformBuffer = scaledEnvelopeWaveform.Data.GetWritableBuffer();
+
+            // populate cloned waveform with scaled waveform data
+            for (int i = 0; i < envelope.Length; i++)
+                scaledEnvelopeWaveformBuffer[i] = ComplexSingle.FromSingle(envelope[i]);
+
+            // get statistics on the waveform
+            float max = envelope.Max();
+            float min = envelope.Min();
+            float gain = (max - min) * 0.5f; // distance between middle of waveform and max/min
+            float offset = min + gain;
+            float absolutePeak = Math.Abs(offset) + gain; // peak voltage of the waveform with absolute value of offset
+
+            // scale waveform to peak voltage
+            for (int i = 0; i < envelope.Length; i++)
+                envelope[i] = envelope[i] / (absolutePeak); // brings waveform down to +/- 1 magnitude
+
+            // set instrument properties
+            envVsg.IQOutPort[""].Level = 2 * absolutePeak; // gain is interpreted as peak-to-peak
+            envVsg.IQOutPort[""].Offset = 0.0; // set offset to 0 since this is done in DSP not in HW on the 5820 and only clips the waveform further
+
+            // create another waveform that we can use to download the scaled envelope to the instrument
+            Waveform instrEnvelopeWaveform = envelopeWaveform;
+            instrEnvelopeWaveform.Data = envelopeWaveform.Data.Clone();
+            WritableBuffer<ComplexSingle> instrEnvelopeWaveformBuffer = scaledEnvelopeWaveform.Data.GetWritableBuffer();
+
+            // populate cloned waveform with scaled waveform data
+            for (int i = 0; i < envelope.Length; i++)
+                instrEnvelopeWaveformBuffer[i] = ComplexSingle.FromSingle(envelope[i]);
+
+            SG.DownloadWaveform(envVsg, instrEnvelopeWaveform); // download optimized waveform
+
+            return scaledEnvelopeWaveform; // return the waveform as it will appear coming out of the front end of the envelope generator
+        }
+
+        public static void InitiateSynchronousGeneration(NIRfsg rfVsg, NIRfsg envVsg, SynchronizationConfiguration syncConfig)
+        {   
+            TClock tclk = new TClock(new ITClockSynchronizableDevice[] { rfVsg, envVsg });
+            tclk.DevicesToSynchronize[0].SampleClockDelay = -syncConfig.RFDelayRange_s / 2.0; // The VST2 can only apply positive delays so we have to establish an inital delay of -RFDelayRange/2 for TCLK to handle negative shifts as well
+            rfVsg.Arb.RelativeDelay = syncConfig.RFDelayRange_s + syncConfig.RFDelay_s;
+            tclk.ConfigureForHomogeneousTriggers();
+            tclk.Synchronize(); // #todo: allow drift?
+            tclk.Initiate();
+        }
+        #endregion
+
+        private static float[] LinearInterpolation1D(float[] x, float[] y, float[] xi, bool monotonic = false)

```

Comment:
I'll just have to take your word for it on this one. Maybe some unit tests to validate?

Reply:
2a092eea30e338dbc381b8f08f19307d176c303f